### ISSUE TITLE
Connectivity timeout after auth/proxy rotation

### ISSUE THREAD

**Issue by Aharic** [gravitational/teleport#12063]
### Description

**What happened**:
Customer environment requires frequent auth/proxy rotations and scale-up/scale-down. Whenever auth and proxy are rotated, however, customer observes that they're not able to connect to nodes for 10 or so minutes after. Customer has upgraded to 9.0.3 on all components to take advantage of the tunnel and auth stability fixes, but still continues to see the issue. Logs from node and proxy taken at the same time are below.

### Server Details
- Teleport version (run `teleport version`): 9.0.3

### Debug Logs

Node:

```
Apr 6, 2022 @ 16:12:41.460	INFO [PROC:1] Teleport component "node" has recovered from a degraded state. service/state.go:123
	Apr 6, 2022 @ 16:12:26.469	INFO [PROC:1] Teleport component "node" is recovering from a degraded state. service/state.go:119
	Apr 6, 2022 @ 16:12:21.461	INFO [PROC:1] Detected Teleport component "node" is running in a degraded state. service/state.go:105
	Apr 6, 2022 @ 16:12:21.461	WARN [NODE:BEAT] Keep alive has failed: error reading from server: EOF. srv/heartbeat.go:584
	Apr 6, 2022 @ 16:12:21.461	WARN [NODE:BEAT] Heartbeat failed keep alive channel closed. srv/heartbeat.go:273
	Apr 6, 2022 @ 16:11:23.911	WARN [NODE:1] Restart watch on error: watcher is closed. resource-kind:lock services/watcher.go:218
	Apr 6, 2022 @ 16:11:21.464	ERRO [NODE] error reading from server: EOF regular/sshserver.go:1774
	Apr 6, 2022 @ 16:11:21.464	WARN [PROC:1] Sync rotation state cycle failed. Retrying in ~10s service/connect.go:487
	Apr 6, 2022 @ 16:11:21.464	INFO [AUDIT] session.data addr.remote:10.104.3.38:33806 code:T2006I ei:2.147483646e+09 event:session.data login:ubuntu namespace:default rx:5014 server_id:538639307912-i-00438a1cb07703db2 sid: time:2022-04-06T14:11:21.463Z tx:6761 uid:ec891522-dea8-453f-8108-25ae7ac3d69e user:example.user@domain.com events/emitter.go:325
	Apr 6, 2022 @ 16:11:21.464	WARN [NODE:1:CA] Re-init the cache on error: watcher is closed: error reading from server: EOF. cache/cache.go:791
	Apr 6, 2022 @ 16:09:44.204	INFO [SESSION:N] New party ServerContext(10.104.3.38:55038->10.102.33.0:52624, user=ubuntu, id=5) party(id=441001f9-009a-404b-8aab-ff85746bf587) joined session: 2fd86a49-5589-4975-9d7a-b24ceb4f433c srv/sess.go:1516
	Apr 6, 2022 @ 16:09:44.184	INFO [NODE] Creating (interactive) session 2fd86a49-5589-4975-9d7a-b24ceb4f433c. cert:2022-04-07 19:52:37 +0000 UTC id:5 idle:30m0s local:10.102.33.0:52624 login:ubuntu remote:10.104.3.38:55038 teleportUser:example.user@domain.com srv/sess.go:223
```

Proxy Logs:
```
Apr 6, 2022 @ 16:12:54.122	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:12:50.715	message repeated 43 times: [ WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559]
	Apr 6, 2022 @ 16:12:18.718	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:12:17.699	#011/go/src/github.com/gravitational/teleport/lib/web/terminal.go:235 github.com/gravitational/teleport/lib/web.(*TerminalHandler).handler
	Apr 6, 2022 @ 16:12:17.699	#011/go/src/github.com/gravitational/teleport/lib/httplib/httplib.go:67 github.com/gravitational/teleport/lib/httplib.MakeHandlerWithErrorWriter.func1
	Apr 6, 2022 @ 16:12:17.699	#011/go/src/github.com/gravitational/teleport/lib/web/apiserver.go:451 github.com/gravitational/teleport/lib/web.NewHandler.func1
	Apr 6, 2022 @ 16:12:17.699	#011/tmp/gomodcache/github.com/gravitational/oxy@v0.0.0-20211213172937-a1ba0900a4c9/ratelimit/tokenlimiter.go:118 github.com/gravitational/oxy/ratelimit.(*TokenLimiter).ServeHTTP
	Apr 6, 2022 @ 16:12:17.699	#011/opt/go/src/runtime/asm_amd64.s:1581 runtime.goexit
	Apr 6, 2022 @ 16:12:17.699	#011/go/src/github.com/gravitational/teleport/lib/web/apiserver.go:1912 github.com/gravitational/teleport/lib/web.(*Handler).siteNodeConnect
	Apr 6, 2022 @ 16:12:17.699	#011/opt/go/src/net/http/server.go:2090 net/http.StripPrefix.func1
	Apr 6, 2022 @ 16:12:17.699	#011/tmp/gomodcache/github.com/julienschmidt/httprouter@v1.3.0/router.go:448 github.com/julienschmidt/httprouter.(*Router).ServeHTTP
	Apr 6, 2022 @ 16:12:17.699	#011/opt/go/src/net/http/server.go:2879 net/http.serverHandler.ServeHTTP
	Apr 6, 2022 @ 16:12:17.699	#011/go/src/github.com/gravitational/teleport/lib/web/terminal.go:206 github.com/gravitational/teleport/lib/web.(*TerminalHandler).Serve
	Apr 6, 2022 @ 16:12:17.699	#011/tmp/gomodcache/github.com/julienschmidt/httprouter@v1.3.0/router.go:387 github.com/julienschmidt/httprouter.(*Router).ServeHTTP
	Apr 6, 2022 @ 16:12:17.699	#011/opt/go/src/net/http/server.go:2047 net/http.HandlerFunc.ServeHTTP
	Apr 6, 2022 @ 16:12:17.699	#011/tmp/gomodcache/github.com/gravitational/oxy@v0.0.0-20211213172937-a1ba0900a4c9/connlimit/connlimit.go:75 github.com/gravitational/oxy/connlimit.(*ConnLimiter).ServeHTTP
	Apr 6, 2022 @ 16:12:17.699	User Message: connection error: desc = &#34;transport: authentication handshake failed: context deadline exceeded&#34;] web/terminal.go:237
	Apr 6, 2022 @ 16:12:17.699	#011/go/src/github.com/gravitational/teleport/lib/web/terminal.go:301 github.com/gravitational/teleport/lib/web.(*TerminalHandler).makeClient
	Apr 6, 2022 @ 16:12:17.699	#011/go/src/github.com/gravitational/teleport/lib/web/apiserver.go:2496 github.com/gravitational/teleport/lib/web.(*Handler).WithClusterAuth.func1
	Apr 6, 2022 @ 16:12:17.699	#011/opt/go/src/net/http/server.go:2047 net/http.HandlerFunc.ServeHTTP
	Apr 6, 2022 @ 16:12:17.699	#011/go/src/github.com/gravitational/teleport/lib/web/apiserver.go:215 github.com/gravitational/teleport/lib/web.(*APIHandler).ServeHTTP
	Apr 6, 2022 @ 16:12:17.699	#011/opt/go/src/net/http/server.go:1930 net/http.(*conn).serve
	Apr 6, 2022 @ 16:12:17.698	INFO [WEBSOCKET] Failed creating a client for session 0f08c7dd-8797-4183-b6cc-41e9a502d381. error:[
	Apr 6, 2022 @ 16:12:17.698	ERROR REPORT:
	Apr 6, 2022 @ 16:12:17.698	Original Error: *trace.ConnectionProblemError connection error: desc = &#34;transport: authentication handshake failed: context deadline exceeded&#34;
	Apr 6, 2022 @ 16:12:17.698	Stack Trace:
	Apr 6, 2022 @ 16:12:17.698	#011/go/src/github.com/gravitational/teleport/api/client/client.go:1379 github.com/gravitational/teleport/api/client.(*Client).IsMFARequired
	Apr 6, 2022 @ 16:12:17.698	#011/go/src/github.com/gravitational/teleport/lib/web/terminal.go:320 github.com/gravitational/teleport/lib/web.(*TerminalHandler).issueSessionMFACerts
	Apr 6, 2022 @ 16:12:17.698	#011/go/src/github.com/gravitational/teleport/lib/client/client.go:363 github.com/gravitational/teleport/lib/client.(*ProxyClient).IssueUserCertsWithMFA
	Apr 6, 2022 @ 16:12:15.082	message repeated 11 times: [ WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559]
	Apr 6, 2022 @ 16:11:57.903	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:11:57.696	INFO [CLIENT] Successful auth with proxy localhost:3080. client/api.go:2299
	Apr 6, 2022 @ 16:11:57.685	INFO [CLIENT] Connecting to proxy=localhost:3080 login="ubuntu" using TLS Routing client/api.go:2294
	Apr 6, 2022 @ 16:11:57.682	INFO [WEB] Getting terminal to &web.TerminalRequest{Server:"538639307912-i-00438a1cb07703db2", Login:"ubuntu", Term:session.TerminalParams{W:317, H:76}, SessionID:"0f08c7dd-8797-4183-b6cc-41e9a502d381", Namespace:"default", ProxyHostPort:"localhost:3080,3023", Cluster:"teleport-dev", InteractiveCommand:[]string(nil), KeepAliveInterval:300000000000}. web/apiserver.go:1911
	Apr 6, 2022 @ 16:11:57.538	message repeated 15 times: [ WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559]
	Apr 6, 2022 @ 16:11:47.512	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:11:45.979	WARN [WEBSOCKET] Unable to stream terminal: connection error: desc = "transport: authentication handshake failed: context deadline exceeded". web/terminal.go:434
	Apr 6, 2022 @ 16:11:44.998	message repeated 4 times: [ WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559]
	Apr 6, 2022 @ 16:11:34.975	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:11:33.976	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:11:33.951	ERROR REPORT:
	Apr 6, 2022 @ 16:11:33.951	User Message: remote error: tls: bad certificate] alpnproxy/proxy.go:290
	Apr 6, 2022 @ 16:11:33.951	User Message: remote error: tls: bad certificate] alpnproxy/proxy.go:290
	Apr 6, 2022 @ 16:11:33.951	Original Error: *tls.permanentError remote error: tls: bad certificate
	Apr 6, 2022 @ 16:11:33.951	Stack Trace:
	Apr 6, 2022 @ 16:11:33.951	#011/go/src/github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:344 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).handleConn
	Apr 6, 2022 @ 16:11:33.951	#011/go/src/github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:282 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).Serve.func1
	Apr 6, 2022 @ 16:11:33.951	#011/go/src/github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:282 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).Serve.func1
	Apr 6, 2022 @ 16:11:33.951	WARN [ALPN:PROX] Failed to handle client connection. error:[
	Apr 6, 2022 @ 16:11:33.951	#011/opt/go/src/runtime/asm_amd64.s:1581 runtime.goexit
	Apr 6, 2022 @ 16:11:33.951	#011/opt/go/src/runtime/asm_amd64.s:1581 runtime.goexit
	Apr 6, 2022 @ 16:11:33.950	Original Error: *tls.permanentError remote error: tls: bad certificate
	Apr 6, 2022 @ 16:11:33.950	Stack Trace:
	Apr 6, 2022 @ 16:11:33.950	WARN [ALPN:PROX] Failed to handle client connection. error:[
	Apr 6, 2022 @ 16:11:33.950	#011/go/src/github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:344 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).handleConn
	Apr 6, 2022 @ 16:11:33.950	ERROR REPORT:
	Apr 6, 2022 @ 16:11:33.898	message repeated 9 times: [ WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559]
	Apr 6, 2022 @ 16:11:27.170	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:11:26.058	message repeated 33 times: [ WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559]
	Apr 6, 2022 @ 16:11:25.923	INFO [CLIENT] Client=10.104.3.38:33852 connecting to node=babel on cluster teleport-dev client/client.go:1074
	Apr 6, 2022 @ 16:11:25.920	INFO [CLIENT] Successful auth with proxy localhost:3080. client/api.go:2299
	Apr 6, 2022 @ 16:11:25.912	INFO [CLIENT] Connecting to proxy=localhost:3080 login="ubuntu" using TLS Routing client/api.go:2294
	Apr 6, 2022 @ 16:11:25.890	INFO [CLIENT] Successful auth with proxy localhost:3080. client/api.go:2299
	Apr 6, 2022 @ 16:11:25.879	INFO [CLIENT] Connecting to proxy=localhost:3080 login="ubuntu" using TLS Routing client/api.go:2294
	Apr 6, 2022 @ 16:11:25.876	INFO [WEB] Getting terminal to &web.TerminalRequest{Server:"538639307912-i-00438a1cb07703db2", Login:"ubuntu", Term:session.TerminalParams{W:317, H:76}, SessionID:"77f889b3-a15c-4cd3-a750-00740e9f9be8", Namespace:"default", ProxyHostPort:"localhost:3080,3023", Cluster:"teleport-dev", InteractiveCommand:[]string(nil), KeepAliveInterval:300000000000}. web/apiserver.go:1911
	Apr 6, 2022 @ 16:11:25.300	message repeated 2 times: [ WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559]
	Apr 6, 2022 @ 16:11:23.705	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:11:21.465	error reading from server: EOF
	Apr 6, 2022 @ 16:11:21.465	WARN [WEBSOCKET] Unable to stream terminal: ssh: could not start shell. web/terminal.go:434
	Apr 6, 2022 @ 16:11:20.730	message repeated 32 times: [ WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559]
	Apr 6, 2022 @ 16:11:05.245	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:11:04.065	message repeated 20 times: [ WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559]
	Apr 6, 2022 @ 16:10:33.741	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:10:32.059	ERROR REPORT:
	Apr 6, 2022 @ 16:10:32.059	Original Error: *net.OpError read tcp 10.104.12.103:3080-&gt;10.102.1.155:52786: read: connection reset by peer
	Apr 6, 2022 @ 16:10:32.059	Stack Trace:
	Apr 6, 2022 @ 16:10:32.059	#011/go/src/github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:344 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).handleConn
	Apr 6, 2022 @ 16:10:32.059	#011/go/src/github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:282 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).Serve.func1
	Apr 6, 2022 @ 16:10:32.059	#011/opt/go/src/runtime/asm_amd64.s:1581 runtime.goexit
	Apr 6, 2022 @ 16:10:32.059	User Message: read tcp 10.104.12.103:3080-&gt;10.102.1.155:52786: read: connection reset by peer] alpnproxy/proxy.go:290
	Apr 6, 2022 @ 16:10:32.058	ERROR REPORT:
	Apr 6, 2022 @ 16:10:32.058	Original Error: *tls.permanentError remote error: tls: bad certificate
	Apr 6, 2022 @ 16:10:32.058	Stack Trace:
	Apr 6, 2022 @ 16:10:32.058	#011/go/src/github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:344 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).handleConn
	Apr 6, 2022 @ 16:10:32.058	#011/go/src/github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:282 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).Serve.func1
	Apr 6, 2022 @ 16:10:32.058	#011/opt/go/src/runtime/asm_amd64.s:1581 runtime.goexit
	Apr 6, 2022 @ 16:10:32.058	User Message: remote error: tls: bad certificate] alpnproxy/proxy.go:290
	Apr 6, 2022 @ 16:10:32.058	WARN [ALPN:PROX] Failed to handle client connection. error:[
	Apr 6, 2022 @ 16:10:32.057	WARN [ALPN:PROX] Failed to handle client connection. error:[
	Apr 6, 2022 @ 16:10:30.621	WARN [PROXY] Restart watch on error: watcher is closed. resource-kind:lock services/watcher.go:218
	Apr 6, 2022 @ 16:10:30.601	message repeated 2 times: [ WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559]
	Apr 6, 2022 @ 16:10:29.597	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:10:29.310	WARN [PROXY:1] Restart watch on error: watcher is closed. resource-kind:proxy services/watcher.go:218
	Apr 6, 2022 @ 16:10:29.112	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:10:28.669	WARN [PROXY] Restart watch on error: watcher is closed. resource-kind:lock services/watcher.go:218
	Apr 6, 2022 @ 16:10:28.374	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:10:28.331	INFO [CLIENT] Client=10.104.3.38:33806 connecting to node=babel on cluster teleport-dev client/client.go:1074
	Apr 6, 2022 @ 16:10:28.328	INFO [CLIENT] Successful auth with proxy localhost:3080. client/api.go:2299
	Apr 6, 2022 @ 16:10:28.319	INFO [CLIENT] Connecting to proxy=localhost:3080 login="ubuntu" using TLS Routing client/api.go:2294
	Apr 6, 2022 @ 16:10:28.301	INFO [CLIENT] Successful auth with proxy localhost:3080. client/api.go:2299
	Apr 6, 2022 @ 16:10:28.289	INFO [CLIENT] Connecting to proxy=localhost:3080 login="ubuntu" using TLS Routing client/api.go:2294
	Apr 6, 2022 @ 16:10:28.286	INFO [WEB] Getting terminal to &web.TerminalRequest{Server:"538639307912-i-00438a1cb07703db2", Login:"ubuntu", Term:session.TerminalParams{W:317, H:76}, SessionID:"27ad9325-58a0-4060-b115-57c17f696d76", Namespace:"default", ProxyHostPort:"localhost:3080,3023", Cluster:"teleport-dev", InteractiveCommand:[]string(nil), KeepAliveInterval:300000000000}. web/apiserver.go:1911
	Apr 6, 2022 @ 16:10:28.252	WARN [PROXY] Restart watch on error: watcher is closed. resource-kind:lock services/watcher.go:218
	Apr 6, 2022 @ 16:10:27.639	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:10:27.627	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:10:27.556	WARN [PROXY:1] Restart watch on error: watcher is closed. resource-kind:proxy services/watcher.go:218
	Apr 6, 2022 @ 16:10:27.297	WARN [PROXY:1] Restart watch on error: watcher is closed. resource-kind:proxy services/watcher.go:218
	Apr 6, 2022 @ 16:10:27.249	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:10:26.980	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:10:26.602	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:10:25.192	WARN [REVERSE:T] Re-init the cache on error: watcher is closed: error reading from server: EOF. cache/cache.go:791
	Apr 6, 2022 @ 16:10:25.191	WARN [PROXY:1:C] Re-init the cache on error: watcher is closed: error reading from server: EOF. cache/cache.go:791
	Apr 6, 2022 @ 16:10:25.191	WARN [PROC:1] Sync rotation state cycle failed. Retrying in ~10s service/connect.go:487
	Apr 6, 2022 @ 16:10:25.190	WARN [PROXY:1:C] Re-init the cache on error: watcher is closed: error reading from server: EOF. cache/cache.go:791
	Apr 6, 2022 @ 16:10:25.190	WARN [REVERSE:T] Re-init the cache on error: watcher is closed: error reading from server: EOF. cache/cache.go:791
	Apr 6, 2022 @ 16:10:25.190	WARN [PROC:1] Sync rotation state cycle failed. Retrying in ~10s service/connect.go:487
	Apr 6, 2022 @ 16:10:25.189	WARN [PROXY:1:C] Re-init the cache on error: watcher is closed: error reading from server: EOF. cache/cache.go:791
	Apr 6, 2022 @ 16:10:25.189	WARN [REVERSE:T] Re-init the cache on error: watcher is closed: error reading from server: EOF. cache/cache.go:791
	Apr 6, 2022 @ 16:10:25.188	WARN [PROC:1] Sync rotation state cycle failed. Retrying in ~10s service/connect.go:487
	Apr 6, 2022 @ 16:10:25.122	message repeated 25 times: [ WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559]
	Apr 6, 2022 @ 16:10:24.771	message repeated 12 times: [ WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559]
	Apr 6, 2022 @ 16:10:23.758	message repeated 16 times: [ WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559]
	Apr 6, 2022 @ 16:09:57.563	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:09:57.537	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:09:57.433	WARN [WEBSOCKET] Unable to send error to terminal: wait: remote command exited without exit status or exit signal: websocket: close sent. web/terminal.go:437
	Apr 6, 2022 @ 16:09:57.433	WARN [WEBSOCKET] Unable to stream terminal: wait: remote command exited without exit status or exit signal. web/terminal.go:434
	Apr 6, 2022 @ 16:09:57.418	ERRO [CLIENT] websocket: close sent client/session.go:655
	Apr 6, 2022 @ 16:09:57.332	WARN [WEBSOCKET] Unable to send error to terminal: wait: remote command exited without exit status or exit signal: websocket: close sent. web/terminal.go:437
	Apr 6, 2022 @ 16:09:57.332	WARN [WEBSOCKET] Unable to stream terminal: wait: remote command exited without exit status or exit signal. web/terminal.go:434
	Apr 6, 2022 @ 16:09:57.316	ERRO [CLIENT] websocket: close sent client/session.go:655
	Apr 6, 2022 @ 16:09:57.156	message repeated 19 times: [ WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559]
	Apr 6, 2022 @ 16:09:56.306	message repeated 122 times: [ WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559]
	Apr 6, 2022 @ 16:09:44.640	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:09:44.075	INFO [CLIENT] Client=10.104.3.38:55038 connecting to node=babel on cluster teleport-dev client/client.go:1074
	Apr 6, 2022 @ 16:09:44.072	INFO [CLIENT] Successful auth with proxy localhost:3080. client/api.go:2299
	Apr 6, 2022 @ 16:09:44.064	INFO [CLIENT] Connecting to proxy=localhost:3080 login="ubuntu" using TLS Routing client/api.go:2294
	Apr 6, 2022 @ 16:09:44.044	INFO [CLIENT] Successful auth with proxy localhost:3080. client/api.go:2299
	Apr 6, 2022 @ 16:09:44.032	INFO [CLIENT] Connecting to proxy=localhost:3080 login="ubuntu" using TLS Routing client/api.go:2294
	Apr 6, 2022 @ 16:09:44.029	INFO [WEB] Getting terminal to &web.TerminalRequest{Server:"538639307912-i-00438a1cb07703db2", Login:"ubuntu", Term:session.TerminalParams{W:317, H:76}, SessionID:"2fd86a49-5589-4975-9d7a-b24ceb4f433c", Namespace:"default", ProxyHostPort:"localhost:3080,3023", Cluster:"teleport-dev", InteractiveCommand:[]string(nil), KeepAliveInterval:300000000000}. web/apiserver.go:1911
	Apr 6, 2022 @ 16:09:43.653	message repeated 82 times: [ WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559]
	Apr 6, 2022 @ 16:09:18.927	WARN [WEB] http: named cookie not present request:GET /web/launch/teleport.cluster.local web/apiserver.go:2559
	Apr 6, 2022 @ 16:09:18.038	INFO [CLIENT] Client=10.104.3.38:55010 connecting to node=babel on cluster teleport-dev client/client.go:1074
	Apr 6, 2022 @ 16:09:18.031	INFO [CLIENT] Successful auth with proxy localhost:3080. client/api.go:2299
	Apr 6, 2022 @ 16:09:18.023	INFO [CLIENT] Connecting to proxy=localhost:3080 login="ubuntu" using TLS Routing client/api.go:2294
	Apr 6, 2022 @ 16:09:17.997	INFO [CLIENT] Successful auth with proxy localhost:3080. client/api.go:2299
```

gz#4761

**Comment by russjones** [comment]
@Aharic How many nodes are in this cluster?

**Comment by russjones** [comment]
This initially looked like https://github.com/gravitational/teleport/issues/10019 which was fixed in https://github.com/gravitational/teleport/pull/11229, but the error message is slightly different. That one was:

```
authentication handshake failed: read tcp 192.168.248.193:65080-&gt;172.31.150.10:443: i/o timeout&#34
```

Here it is:

```
transport: authentication handshake failed: context deadline exceeded&#34
```

**Comment by Aharic** [comment]
@russjones ,

Customer confirmed they're running around 200 nodes total. Systemd config file reprinted below:

```
[Unit]
Description=Teleport Auth Service
After=network.target

[Service]
Type=simple
Restart=on-failure
ExecStart=/usr/local/bin/teleport start --roles=auth --config=/etc/teleport.yaml --diag-addr=0.0.0.0:3434	 --pid-file=/run/teleport.pid
ExecReload=/bin/kill -HUP $MAINPID
PIDFile=/run/teleport.pid
LimitNOFILE=65536
StandardOutput=syslog
StandardError=syslog
SyslogIdentifier=teleport-auth

[Install]
WantedBy=multi-user.target
```

**Comment by russjones** [comment]
@espadolini It looks like the customer is sending a HUP to Teleport starting graceful restart. With the issues you know that currently exist with graceful restart, does it line up that you could see a 10-20 minute outage?

**Comment by espadolini** [comment]
From a quick check it seems like we are shutting down the `regular.Server` before closing the `agentPool` in `ssh.shutdown`, so if a node that's shutting down is keeping a connection up to the new proxy then until all connections close on that node, it won't answer to new connections? I'll investigate.

**Comment by espadolini** [comment]
@Aharic was this issue encountered during a rotation? Are they using reverse tunnel nodes, or direct dial?

**Comment by russjones** [comment]
> @russjones you mean https://github.com/gravitational/teleport/issues/12031?

Yes, sorry updated.

**Comment by Aharic** [comment]
@espadolini, 

These are the answers from the customer:

-> During rotation

-> We are using Reverse tunnel

-> For certain period, none of the nodes are accessible at all until everything works again perfectly after certain period time.

**Comment by russjones** [comment]
@Aharic When you say rotation, do you mean CA rotation?

**Comment by Aharic** [comment]
@russjones should have been more clear. they mean during pod rotation during upgrade. @espadolini , we've requested the info from the customer and I'll update re: specific log when they respond. So far they've confirmed that they can log in normally to webui and tsh, they just can't connect to anything. Customer has also noted that they're happy to jump on a live call with our dev team if that would help narrow down the issue or gather the required information. If that's something the dev team is open to @pschisa and I can schedule a call with the customer for a time that works for all parties.

**Comment by pschisa** [comment]
@espadolini just following up to see if you need any additional information.  This is currently in a test environment so the customer is happy to meet up / test code as needed to help resolve the issue

**Comment by espadolini** [comment]
@pschisa I think I have found a possible cause but I'm still trying to figure out what triggers it; once I know what to look for I'll reach out so we can try replicating it in the customer's test env as well.

Can you confirm that they are spinning up new auth/proxy combo pods before gracefully shutting down the old ones? (a rolling upgrade of the deployment would do that)

**Comment by pschisa** [comment]
Apologies that this wasnt clear before but they use VMs and this is when they resize their auth servers specifically (they are separate from the proxies).  Do you still want me to confirm if they do rolling upgrades of the auth VM?

**Comment by Aharic** [comment]
@espadolini , customer confirmed that it stops working when they shrink the group down to do the upgrades. They use [spot.io](http://spot.io/) elastigroups but they seem to work similar to autoscale groups. Are we saying that customer has to shut down ALL auth instances and bring the cluster down completely in order to run an upgrade? We've always been under the impression that scaling down to a single node/pod/instance, upgrading, and then scaling back up was supported.

**Comment by espadolini** [comment]
The cluster at large doesn't have to go down, you can keep older components running, but it's important that the only auth running is the one instance running the new version, to ensure that upgrades work smoothly (especially important when upgrading through major versions). Everything else should be able to handle the auth being down for a minute or two, and reconnect gracefully. The rest of the cluster can be upgraded afterwards.

**Comment by pschisa** [comment]
@espadolini Latest from the customer, seems they are following the shrink down to 1 rule:

"If we shrink the group down to 1 and attempt to resize that way, this is when we see the complete outage. If we do a rolling upgrade, however, where the instances are resized 1 by 1, it's only a partial unavailability"

Let me know if any further information can be provided, they are happy to review the issue live and load debugging builds if needed

**Comment by zmb3** [comment]
These issues should be resolved by:

- #12488
- #12540
- #12557

**Comment by espadolini** [comment]
Yes, plus at least one fix from #12359 in the agentpool code (as proxy peering is going to be v9+ only).

**Comment by programmerq** [comment]
@pschisa Your stacktrace has `read: connection reset by peer` near the top. Double check that there are no disagreements with the keepalive frequency and the timeout duration on any load balancers. Usually an error like `read: connection reset by peer` indicates that something actively closes the connection.

A telltale sign that is happening is if you grab a packet capture with tshark/tcpdump and load it up in wireshark. a TCP RST packet can be sent by a load balancer or firewall to both the client and server to forcefully close a connection like this. If the TTL of the TCP RST packet isn't the same as other packets in the stream.

**Comment by espadolini** [comment]
The weird behavior was caused by a logic error in how the proxy handled connections to the auth on behalf of clients - especially nodes, in this situation: instead of forwarding connections to the same load balancer that was used by the proxy itself to connect to auth, connections were forwarded to individual auth servers as stored in the backend heartbeat resources, with serialized connection attempts in order of auth UUID in most cases (in the case of ALPN `teleport-auth` connections, as used by tsh with TLS routing enabled, in random order).

This went unnoticed for this long (the `RemoteAuthServer` code in the reversetunnel `transport` handler is years old at this point) because most HA setups end up using internal kubernetes networking, which will fail connections pretty quickly if the destination address doesn't exist anymore, so the proxy quickly cycles through all the addresses in the backend to land on a working one - but the user experiencing this problem uses full VMs with a real network between them, so connections from nodes through the proxy to the auth after a downscale were sent to nowhere, waiting for a TCP timeout that was longer than the connection timeout on the node itself, so any Teleport component attempting to connect to an auth server through the proxy failed to do so until all the heartbeat resources of the auth servers that were downscaled ended up expiring from the backend.

The problem was successfully worked around by setting `advertise_ip` in the auth configuration to point to the auth load balancer instead (which makes every auth in the backend pretend to be directly accessible through the load balancer address, thus restoring the correct behavior), and should be fixed by #13310, with the v9 backport #13399 released as part of [v9.3.5](https://github.com/gravitational/teleport/releases/v9.3.5).

**Comment by espadolini** [comment]
User confirmed that v9.3.5 fixed the issue without having to use the `advertise_ip` workaround.
